toolkit/tools/imagegen/diskutils/initramfs.go

```diff
@@ -0,0 +1,229 @@
+	// Caller must use InitramfsMount.Close() to clean up the outputs
+	initramfs.outputBuffer = new(bytes.Buffer)
+	initramfs.pgzWriter = pgzip.NewWriter(initramfs.outputBuffer)
+	initramfs.cpioWriter = cpio.NewWriter(initramfs.pgzWriter)
+
+	initramfs.initramfsOutputFile, err = os.OpenFile(initramfsPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, initramfsModeBits)
+
+	return
+}
+
+// OpenInitramfs makes an existing initramfs editable
+func OpenInitramfs(initramfsPath string) (initramfs InitramfsMount, err error) {
+	logger.Log.Debugf("Opening intramfs '%s'", initramfsPath)
+	inputFile, err := os.Open(initramfsPath)
+	if err != nil {
+		return
+	}
+	defer inputFile.Close()
+
+	gzReader, err := pgzip.NewReader(inputFile)
+	if err != nil {
+		return
+	}
+	defer gzReader.Close()
+	cpioReader := cpio.NewReader(gzReader)
+	//cpio.Reader has no Close() function to defer
+
+	// Caller must use InitramfsMount.Close() to clean up the outputs
+	initramfs.outputBuffer = new(bytes.Buffer)
+	initramfs.pgzWriter = pgzip.NewWriter(initramfs.outputBuffer)
+	initramfs.cpioWriter = cpio.NewWriter(initramfs.pgzWriter)
+
+	var bytesIO int64
+	var nextFileHeader *cpio.Header
+	for {
+		var linkPayload []byte
+
+		nextFileHeader, err = cpioReader.Next()
+		if err == io.EOF {
+			err = nil
+			break
+		}
+		if err != nil {
+			return
+		}
+
+		// For a given symlink, generate a payload that contains the read value of the link.
+		// The payload should be written after the header.
+		// e.g. A link from (/bin -> /usr/bin) would have a payload of "/usr/bin").
+		// cpio.ModeSymlink is two bits (one of which is resused), need to check for actual
+		// equality rather than non-zero after masking
+		isLink := (nextFileHeader.Mode&cpio.ModeType == cpio.ModeSymlink)
+		if isLink {
+			linkPayload = []byte(nextFileHeader.Linkname)
+			nextFileHeader.Size = int64(len(linkPayload))
+		}
+
+		err = initramfs.cpioWriter.WriteHeader(nextFileHeader)
+		if err != nil {
+			return
+		}
+
+		if isLink {
+			var bytesWrittenInt int
+
+			// Write returns an int, cast it to an int64 afterwards
+			logger.Log.Tracef("Creating link %s -> %s", nextFileHeader.Name, nextFileHeader.Linkname)
+			bytesWrittenInt, err = initramfs.cpioWriter.Write(linkPayload)
+			bytesIO = int64(bytesWrittenInt)
+
+		} else {
+			bytesIO, err = io.Copy(initramfs.cpioWriter, cpioReader)
+		}
+
+		if err != nil {
+			return
+		}
+
+		logger.Log.Tracef("File %s caused %d bytes to be transfered to new archive", nextFileHeader.Name, bytesIO)
+		logger.Log.Tracef("Buffer unread length: %d", initramfs.outputBuffer.Len())
+	}
+
+	inputFile.Close()
+
+	fileInfo, err := os.Stat(initramfsPath)
+	if err != nil {
+		return
+	}
+
+	// We can't edit a CPIO archive in place, completely overwrite the file with truncate
+	// The output buffer in memory will be used to re-create the initramfs.
+	initramfs.initramfsOutputFile, err = os.OpenFile(initramfsPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileInfo.Mode())
```

Comment:
>os.OpenFile(initramfsPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileInfo.Mode()) [](start = 38, length = 77)

Could we use the output of this `OpenFile` call as the value for `inputFile`, assign `inputFile` here and not close it a few lines above? In case of errors we could check for them in the deferred call and close the file then.